### ISSUE TITLE
Can't change MFA Store config

### ISSUE THREAD

**Issue by palanjam** [neos-sdi/adfsmfa#109]
I'm setting up a brand new install and I can't change the MFAStore config either in MMC (through the Active Directory configuration) or using the powershell Set-MFAStore -Store ADDS -Config $c.

When I try to run the powershell command, I get:

> PS C:\> Set-MFAStore -Store ADDS -Config $c
> 
> Confirm
> Are you sure you want to perform this action?
> Performing the operation "Set-MFAStore" on target "MFA Store Configuration".
> [Y] Yes  [A] Yes to All  [N] No  [L] No to All  [S] Suspend  [?] Help (default is "Y"):
> Set-MFAStore : The Specified directory object cannot be found.
> At line:1 char:1
> + Set-MFAStore -Store ADDS -Config $c
> + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : OperationStopped: (MFA.SetMFAStore:SetMFAStore) [Set-MFAStore], ActiveDirectoryObjectNot
>    FoundException
>     + FullyQualifiedErrorId : 3024,MFA.SetMFAStore

When I try to open the Active Directory configuration folder in MMC, the snap-in crashes with this stack trace:

`at System.DirectoryServices.ActiveDirectory.ActiveDirectorySchemaProperty..ctor(DirectoryContext context, String ldapDisplayName, DirectoryEntry propertyEntry, DirectoryEntry schemaEntry)
   at System.DirectoryServices.ActiveDirectory.ActiveDirectorySchemaProperty.FindByName(DirectoryContext context, String ldapDisplayName)
   at Neos.IdentityServer.MultiFactor.Data.ADDSDataRepositoryService.CheckAttribute(String domainname, String username, String password, String attributename, Int32 multivalued)
   at Neos.IdentityServer.MultiFactor.RuntimeRepository.CheckADDSAttribute(MFAConfig config, String domainname, String username, String password, String attributename, Int32 multivalued)
   at Neos.IdentityServer.MultiFactor.Administration.ManagementService.CheckADDSAttribute(String domainname, String username, String password, String attributename, Int32 multivalued)
   at Neos.IdentityServer.Console.Controls.ADDSConfigurationControl.ValidateData()
   at Neos.IdentityServer.Console.Controls.ADDSConfigurationControl.DoCreateControls()
   at Neos.IdentityServer.Console.Controls.ADDSConfigurationControl.OnCreateControl()
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl()
   at System.Windows.Forms.Control.ControlCollection.Add(Control value)
   at Microsoft.ManagementConsole.View.HandleInitializationRequest(IRequestStatus requestStatus)
   at Microsoft.ManagementConsole.View.ProcessRequest(Request request)
   at Microsoft.ManagementConsole.ViewMessageClient.ProcessRequest(Request request)
   at Microsoft.ManagementConsole.Internal.IMessageClient.ProcessRequest(Request request)
   at Microsoft.ManagementConsole.Executive.RequestStatus.BeginRequest(IMessageClient messageClient, RequestInfo requestInfo)
   at Microsoft.ManagementConsole.Executive.SnapInRequestOperation.ProcessRequest()
   at Microsoft.ManagementConsole.Executive.Operation.OnThreadTransfer(SimpleOperationCallback callback)`

I'm running Windows Server 2012 R2.
I tried this before and after running the schema changes to AD

**Comment by redhook62** [comment]
Hi @palanjam 

It seems that one of the adds attributes that you specified is not, you should be able to correct it with the command PS **Set-MFAStore**, the error message is correct.

You can also use **Set-MFAActiveDirectoryTemplate**.

You can also use Set-MFAActiveDirectoryTemplate, or export the configuration using **Export-MFASystemConfiguration**, edit the resulting xml file, then re-import it using **Import-MFASystemConfiguration**

Be careful, some attributes must be multivalued like **PublicKeyCredentialAttribute**.

However, MMC should not crash, there can be a problem on that side.

It will not be possible to ship a new release before the beginning of August.

Regards

**Comment by palanjam** [comment]
I tried to do the Set-MFAStore with just the defaults and it still gave the same error. I haven't been able to change any of the attributes because of the error I was getting. It just has the defaults in from the initial install (all the msDS-cloudExtension*). 

What does the Set-MFAActiveDirectoryTemplate do? It isn't clear in the wiki. I tried it on a test server and it broke adfsmfa even more. Had to uninstall and redo. 

Both MMC and powershell errors happen straight away even after a fresh install of the msi file and Register-MFAServer.

**Comment by palanjam** [comment]
@redhook62 Could it be that the ClientCertificateAttribute is missing from MMC? Although it doesn't make sense why Powershell wouldn't work since the attribute is there when you Get-MFAStore

**Comment by palanjam** [comment]
I also tried to do the Export-MFASystemConfiguration, edit the file manually and Import it and I get the same powershell error
`Import-MFASystemConfiguration : The specified directory object cannot be found.`

**Comment by 0219dives** [comment]
Hi, I also seem to have an issue using this, I get the error 'MFA Error MFA not initialized ! you must run "Register-MFASystem"' 
This happens when I try and open the MFA MMC.

When I run the command 'Register-MFASystem' I get the following error:
Register-MFASystem : An authentication provider with identifier 'MultiFactorAuthenticationProvider' is already present
in the policy store.  Identifiers must be unique.
+ ~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (MFA.RegisterMFASystem:RegisterMFASystem) [Register-MFASystem], Cmdlet
   InvocationException
    + FullyQualifiedErrorId : 3012,MFA.RegisterMFASystem

For info I'm running this on Server 2012 R2, and hoping to use a SQL Database Store.

Would someone be able to give me some help. 

Thanks so much :)

Dan

**Comment by redhook62** [comment]
> I also tried to do the Export-MFASystemConfiguration, edit the file manually and Import it and I get the same powershell error
> `Import-MFASystemConfiguration : The specified directory object cannot be found.`

Hi @palanjam, @0219dives 

I assure them that using ADFS 2012r2 is not the problem. I do not know what is wrong with your configuration, but when loading the configuration can not be properly deserialized.
The easiest way is to start a new config.

First you can use UnRegister-MFASystem.
If this does not work, refer to the paragraph [Unregister MFA](https://github.com/neos-sdi/adfsmfa/wiki/03-UpgradeFromPrevious#unregister-adfsmfa-authentication-provider-powershell)
Then start a new config.

I think that for ADDS, you have a problem of rights (read / write) on your directory. please check the prerequisites indicated.
Then, if you wish, you can switch to SQL mode.

**Comment by 0219dives** [comment]
Thank you @redhook62  I will give that a go tomorrow! Thanks once again for your quick reply. 👍

**Comment by 0219dives** [comment]
No luck I'm afraid, same issue when I Register-MFASystem it says its already registered, still does not recognise that. Also modified directory permissions to Everyone Full Control however no difference.  

PS C:\Windows\system32> UnRegister-MFASystem

UnRegister-MFASystem : MFA not initialized ! you must run "Register-MFASystem" PowerShell  Cmdlet
At line:1 char:1
+ UnRegister-MFASystem
+ ~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (MFA.UnRegisterMFASystem:UnRegisterMFASystem) [Unregister-MFASystem], Exception
    + FullyQualifiedErrorId : 3013,MFA.UnRegisterMFASystem
 

PS C:\Windows\system32> Register-MFASystem

Register-MFASystem : An authentication provider with identifier 'MultiFactorAuthenticationProvider' is already present in the policy 
store.  Identifiers must be unique.
At line:1 char:1
+ Register-MFASystem
+ ~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (MFA.RegisterMFASystem:RegisterMFASystem) [Register-MFASystem], CmdletInvocationExce 
   ption
    + FullyQualifiedErrorId : 3012,MFA.RegisterMFASystem

**Comment by redhook62** [comment]
@0219dives 

You can do 2 things
- Delete all files found in the ProgarmFiles\MFA\Config (config.db)
- Remove the provider with standard powershell commands

$prov = (Get-AdfsGlobalAuthenticationPolicy).AdditionalAuthenticationProvider

Set-AdfsGlobalAuthenticationPolicy -AdditionalAuthenticationProvider $null

UnRegister-AdfsAuthenticationProvider -Name "MultifactorAuthenticationProvider" -Confirm:$false

At this point the MFA provider is completely unconfigured.
And now, you can start a new configuration

Regards

**Comment by 0219dives** [comment]
Thanks @redhook62 

I've followed those instructions and it removed.

I then ran Register-MFASystem again and got a message 'MFA System correctly registered !' however when I open the MFA MMC I am still getting the error MFA not initialized ! you must run "Register-MFASystem" PowerShell  Cmdlet.

Any ideas?

Thanks,

Dan

**Comment by redhook62** [comment]
Hi, @0219dives 

Now, leaving the MMC aside (we will see if there is a problem with it)
Can you export the configuration using Export-MFASystemConfiguration, and send this file to me at adfsmfa@gmail.com (change any personal information first)
Then you have to try to work with PS.
- Switch to SQL mode

$c = Get-MFAStore SQL
$c.Active = $ true
Set-MFAStore -Kind SQL $c

And check that it goes well.

If everything is OK, you must create an MFA database with the New-MFADatabase cmdlet. (look at the documentation)

Then, test an access to a federated application, just to see if any plant or if you are offered to register, no need to go any further.

Relaunch MMC to see if there has been an impact.

All of these steps will certainly help me understand your concerns.

Let me know your results.

Regards

**Comment by 0219dives** [comment]
Thanks @redhook62 I've just sent over the Config file.

**Comment by redhook62** [comment]
hi **@0219dives , @palanjam** 

After looking at your configuration, the majority of the elements are correct.
The errors in the MMC (this will have to be corrected) seem to come from the fact that the default ADDS schema requires **ADDS 2016.**
To correct this, you have to switch to the "Legacy" template eg: 2003-2012.

To do this in powershell please type the following command: 
**Set-MFAActiveDirectoryTemplate -Kind SchemaAll**

In fact, in the 2016 template there is the **msDS-KeyCredentialLink** attribute which can cause a problem when loading
Then test to see if the MMC works as well.

If no strong authentication method is found you must assign an ADFS policy to your relying party correctly, this does not depend on the component, you must use the ADFS administration console for this.

If all this goes well, each time you save the configuration, a cache file will be created on each ADFS server (ProgramFiles\MFA\Config\config.db)

Thank you for your feedback

redhook

**Comment by 0219dives** [comment]
Thanks redhook.

I've changed that over now.

PS C:\Users\Administrator> Import-MFASystemConfiguration -ImportFilePath C:\system.xml
Confirm
Are you sure you want to perform this action?
Performing the operation "Import and ovveride the current MFA configuration with : C:\system.xml ? " on target
"Import-MFASystemConfiguration".
[Y] Yes  [A] Yes to All  [N] No  [L] No to All  [S] Suspend  [?] Help (default is "Y"): Y
Import-MFASystemConfiguration : The Specified directory object cannot be found.
At line:1 char:1
+ Import-MFASystemConfiguration -ImportFilePath C:\system.xml
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (MFA.ImportMFASystemConfiguration:ImportMFASystemConfiguration) [Impor
   t-MFASystemConfiguration], ActiveDirectoryObjectNotFoundException
    + FullyQualifiedErrorId : 3012,MFA.ImportMFASystemConfiguration

Any ideas?

**Comment by redhook62** [comment]
hi @0219dives , @palanjam 

Good, to short-circuit the loading of the erroneous config, it will be necessary to enter the following commands. but first of all change the PublicKeyCredentialAttribute attribute with the value "otherMailBox" in your config xml file

_Unregister Component_
**$prov = (Get-AdfsGlobalAuthenticationPolicy).AdditionalAuthenticationProvider**

**Set-AdfsGlobalAuthenticationPolicy -AdditionalAuthenticationProvider $null**

**UnRegister-AdfsAuthenticationProvider -Name "MultifactorAuthenticationProvider" -Confirm:$false**


_Register component_
**$ typeName = "Neos.IdentityServer.MultiFactor.AuthenticationProvider, Neos.IdentityServer.MultiFactor, Version = 3.0.0.0, Culture = neutral, PublicKeyToken = 175aa5ee756d2aa2"**

**Register-AdfsAuthenticationProvider -TypeName $ typeName -Name "MultiFactorAuthenticationProvider"
-Verbose -ConfigurationFilePath "c:\temp\config 3.0.xml"**

**net stop adfssrv
net start adfssrv**

**Set-AdfsGlobalAuthenticationPolicy -AdditionalAuthenticationProvider MultiFactorAuthenticationProvider**

After that, in principle things should go back to normal.

We will release in early August to correct the problems you have encountered

Regards

redhook

**Comment by redhook62** [comment]
@0219dives , @palanjam 

I think you are close to the goal.
In federation, it is a usage, the identity of the user MUST be their UserPrincipalName (UPN).
On your relying party add a type claim: http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn 
In your active directory assign aUPN to your users in the form **name@publicdomain.extension**.
this should solve your problems

Regards

redhook

**Comment by 0219dives** [comment]
![Capture](https://user-images.githubusercontent.com/1909433/87977210-477e0d00-cac6-11ea-8c9f-949ab2c42fc0.png)

UPN is in the same format as you mention above, so unsure what is going wrong :(

**Comment by redhook62** [comment]
Are you sure that the upn is not empty ?

**Comment by 0219dives** [comment]
Thanks very much all working! :) I'm very happy!

Unfortunately the MMC isn't working yet, any ideas?

**Comment by redhook62** [comment]
@0219dives 

Very good news !

For MMC, you can try removing the config.db file.
But at the beginning of August, I will do full tests on Windows 2012R2.
Unless you have more informations;

Regards

redhook

**Comment by 0219dives** [comment]
Thanks redhook you have been extreamly helpful! :)

Still no config.db file; only file in that folder is threatconfig.db. 

Any ideas why the config.db file isn't being created?

Thanks,

Dan

**Comment by redhook62** [comment]
Thé config. db file is created each time you save thé config (mmc or ps)  it's cache file for speed.
Remember, export your good config before each modification.
Some things des not work as expected... 

Regards

redhook

**Comment by redhook62** [comment]
You can also try to launch the MMC as Administrator...

**Comment by 0219dives** [comment]
Thanks @redhook62 - Tried running MMC as Administrator but no luck at the moment. 

Thanks,

Dan

**Comment by palanjam** [comment]
I have gotten SQL setup and now I'm able to fully use the MMC. However, I'm now getting an error when I try to use the ADFS login. I'm just using a test server so I have no relying party trusts setup and just trying the login at https://sts.domain.com/adfs/ls/idpinitiatedsignon 
@0219dives What did you do to set it up in ADFS?

**Comment by palanjam** [comment]
Never mind the last comment. I just had to restart the ADFS service to reload the config changes. It's all working now. So to summarize this issue, it appears that the ADDS config is only applicable to Windows 2016. If you're using W2012R2, then the only method that works is the SQL store. Is that accurate @redhook62?

**Comment by 0219dives** [comment]
@palanjam how did you get your mmc working? Are you using server 2016? Thanks, Dan

**Comment by palanjam** [comment]
Nope I'm on 2012 R2. I didn't do anything to get it to work. The only page that was crashing for me was the one where you set all the ADDS settings. But just by clicking on the SQL config and getting that setup got it working like it should. Maybe you have some weird issue with your install? Try completely uninstsllkng it and reinstall

**Comment by redhook62** [comment]
hi @palanjam , @0219dives 

I would like to bring you some informations on the concerns that you both encounter.
First, the component works fine (all modules) with ADFS 2019, 2016 and 2012r2.

Regarding ADDS, the reported problem occurs because in the initial configuration (by default with Register-MFASystem), 
The attributes proposed for the storage of user metadata requires an ADDS 2016 schema (version 85) with in particular a multi-valued attribute "msDS-KeyCredentialLink" (for biometrics) which does not exist in previous versions of ADDS schemas (2003-2012).

The bug is here, when loading the configuration (directly from the ADFS database or from the cache file), 
When you launch the MMC, the Powershell commands or simply in the ADFS portal, a check of the validity of the parameterized attributes is made, an exception is thrown concerning the invalid attribute. normally a warning should be displayed, the rest of the configuration should load without problems. 
As indicated in the comments, we will solve this problem at the beginning of August. We will also pass the ADDS default configuration on the "Legacy" attribute set.

There is a workaround, we must extract the configuration in an xml file, and replace the invalid attribute with a value "otherMailBox".
Then import this file back into ADFS.
You can do this with MFA cmdlets when the key is running ...
**Export-MFASystemConfiguration** and **Import-MFASystemConfiguration** or use ADFS commands
**Export-AdfsAuthenticationProviderConfigurationData** and **Import-AdfsAuthenticationProviderConfigurationData**
Then for an ADDS configuration please respect the prerequisites indicated in the installation part of the wiki

For the console, I do not have specific ideas at the moment. however this is a platform specific concern. the component cannot load its configuration from the ADFS database. as indicated previously this can come from the problem of wrongly defined attributes, but especially from a problem of windows rights. with ADFS 2012R2, it is desirable that the operator is at least Local Administrator of the ADFS server. the AD FS service account should preferably be a domain account

To sum up, on our side.
We will modify the default attribute sets when creating a new configuration, and ensure that in case of loading errors, the MMC and PowerShell commands remain operational.

Best regards

**Comment by redhook62** [comment]
Hi, @palanjam, @0219dives 

Solved in version 3.0.2008.0

Regards
